src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/MethodToStateMachineRewriter.cs

```diff
@@ -863,7 +647,68 @@ public override BoundNode VisitAssignmentOperator(BoundAssignmentOperator node)
 
             // We have an assignment to a variable that has not yet been assigned a proxy.
             // So we assign the proxy before translating the assignment.
-            return HoistRefInitialization(leftLocal, node);
+            var visitedRight = (BoundExpression)Visit(node.Right);
+            return _refInitializationHoister.HoistRefInitialization(
+                leftLocal,
+                visitedRight,
+                proxies,
+                doHoist,
+                createHoistedAccess,
+                this,
+                isRuntimeAsync: false);
+
+            static StateMachineFieldSymbol doHoist(TypeSymbol type, MethodToStateMachineRewriter @this, LocalSymbol assignedLocal)
+            {
+                StateMachineFieldSymbol hoistedSymbol;
+
+                if (@this.F.Compilation.Options.OptimizationLevel == OptimizationLevel.Debug)
```

Comment:
Will we need similar special debug handling for new runtime-async rewriter? If it's possible or likely, please leave a follow-up/tracking link there